### PR TITLE
[None][refactor] Mixed Modality Support for Nemotron Nano Omni V3

### PR DESCRIPTION
**Pull request by aswinvisva**
## Description

Onboards **Nemotron Nano V3 Omni** onto the generic `MultimodalModelMixin` / `EncoderGroup` framework and adds the machinery required for models whose chat template regroups placeholders by modality (Nano's Jinja).

The framework PR ([#16337](https://github.com/NVIDIA/TensorRT-LLM/pull/16337) — "Generic Mixed Modality Support") was previously folded in here; that has since merged to `main` and this branch was rebased on top, so the diff is now scoped strictly to Nemotron + one small framework hook.

### 1. Nemotron Nano V3 Omni migration (`tensorrt_llm/_torch/models/modeling_nemotron_nano.py`)
* `NemotronH_Nano_VL_V2` inherits `MultimodalModelMixin` and registers three per-modality `EncoderGroup`s (`("image",)`, `("video",)`, `("audio",)`) — one group per underlying encoder call (RADIO's image and video sub-paths cannot share a single ViT forward).
* Deletes the old `_encode_multimodal` orchestrator; replaced by `_encode_image_group`, `_encode_video_group`, `_encode_audio_group`. Video-with-embedded-audio interleave stays local to the video group.
* Input processor accepts mixed modality: `_process_images` / `_process_images_dynamic` / `_process_video_prompts` / `_process_audio` now return expanded prompt strings, and `call_with_text_prompt` chains all three and tokenizes once at the end.
* Bucket-key detection replaces `modality_type` reads in `NanoV2VLVisionEncoder.forward`, `apply_evs`, and `_check_encoders_exist`. `modality_type` is preserved for single-modality requests so the untouched EVS-merge path keeps working. Mixed + EVS raises a clear `ValueError` rather than a deep `KeyError`.
* EPD encoder-only wrapper (`NanoV2VLMultimodalEncoder.forward`) routes through `encode_multimodal_by_groups` with per-modality virtual params so the encoder output is modality-major (not input-order), fixing a latent cross-request row-scramble bug on mixed EPD batches.

### 2. Framework hook: `prompt_modality_order` on placeholder metadata
Nano's Jinja regroups placeholders by hardcoded modality order regardless of chat-part send order, so `mm_item_order` (built by chat parsing in send order) disagrees with the actual prompt order. That breaks the framework's `mm_item_order == prompt order` invariant on mixed-modality requests.

Adds a single new optional field on `MultimodalPlaceholderMetadata`:
```python
prompt_modality_order: Optional[Tuple[str, ...]] = None   # e.g. ("image", "video", "audio")
```
`MultimodalDataTracker.item_order()` sorts by this priority when set (per-modality send index as tie break), so the manifest is built in the correct order from the start — no rebuild-then-fix. Default `None` preserves send order for all other models.

Nano declares `prompt_modality_order=("image", "video", "audio")`.

## PR Checklist
- [x] Please check this after reviewing the above items as appropriate for this PR.

## Dev Engineer Review

- Refactors Nemotron NanoV2VL (`NemotronH_Nano_VL_V2`) to inherit from `MultimodalModelMixin` and route multimodal encoding through the generic `EncoderGroup` / `encode_multimodal_by_groups` framework (separate image/video/audio encoder groups, request-order concatenation, and modality-specific grouping/wrapping).
- Updates raw multimodal detection (`has_raw_multimodal_payload`) to consider `multimodal_data["image"|"video"|"audio"]` presence based on any non-`None` values, rather than relying on `modality_type`.
- Adjusts `_lengths_by_modality` enforcement for raw (non-manifest) inputs to derive the request-level modality set by scanning supported raw keys (`image`, `video`, `audio`) present in each `MultimodalParams`; raises when cross-group/multi-modality is detected without `mm_item_order`, and only appends lengths for modalities that exist in the target `by_modality` map.
- Reworks Nemotron NanoV2VL preprocessing/prompt building so multimodal prompts stay as text (`processed_query` remains prompt text) until a single final tokenizer `encode` after assembling all modalities; preserves legacy `multimodal_data["modality_type"]` only for strictly single-modality requests.
- Adds modality-aware prompt ordering support:
  - Introduces `prompt_modality_order` on `MultimodalPlaceholderMetadata`.
  - Adds `MultimodalPlaceholderRegistry.get_prompt_modality_order()`.
  - Updates `MultimodalDataTracker.item_order()` to sort the prompt-order manifest using the configured modality-major order when provided.
- Improves correctness of multimodal routing / EVS:
  - EVS pruning and vision-group sizing are driven by raw `multimodal_data["video"]` metadata (instead of scanning `modality_type` lists).
  - Forward-time validation rejects mixed-modality requests (image/video/audio together) when EVS video pruning is enabled.
  - Encoder-only EPD handoff explicitly rejects any audio found in `param.multimodal_data` to ensure audio is only handled by the audio encoder group (including cases where audio could be extracted from video).
- Unit-test coverage updates include the multimodal caching path: removes older `_encode_multimodal` dispatch/audio-order tests and rewrites caching tests to validate `get_multimodal_embeddings` caching when exercised via the grouped encoder path (`encode_multimodal_by_groups` with per-modality group methods).
- Includes targeted fixes mentioned in the change summary for EVS, multi-image prompts, and mixed EPD batch handling.

## QA Engineer Review

- Test changes (unit tests only):
  - `tests/unittest/_torch/multimodal/test_encoder_group.py`
    - Added `TestLengthsByModality.test_cross_group_multi_modality_without_manifest_raises`
    - Added `TestLengthsByModality.test_single_modality_outside_group_is_skipped`
  - `tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py`
    - Added `TestMmEncoderGroups` and helper functions:
      - `test_property_registers_three_per_modality_groups`
      - `test_image_group_wraps_virtual_params_and_concats`
      - `test_video_group_stashes_num_tokens_in_video`
      - `test_audio_group_concats_encode_audio_outputs`
      - `test_video_group_skips_audio_when_video_has_no_audio`
      - `test_video_group_skips_audio_when_sound_encoder_missing`
      - `test_video_group_interleaves_when_embedded_audio_present`
  - `tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py`
    - Removed `TestEncodeMultimodalDispatch` and `TestEncodeMultimodalAudioOrder`
    - Replaced/rewrote caching/contract coverage with a revised `TestChunkedPrefillCaching` suite for the grouped encoder path.

- Coverage in `test-db/` / `qa/` / manual QA:
  - Not verified from the provided data.
- Verdict: **needs follow-up** (CI previously reported a failing L0 Merge Request pipeline; follow-up CI completion status is not confirmed from the provided context).

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 1]
```diff
@@ -932,23 +939,41 @@ def __init__(self, model_config: ModelConfig[transformers.PretrainedConfig], *ar
...
                 )
 
-        mm_embeddings, _ = super().forward(multimodal_params)
-        if not mm_embeddings:
-            return []
-        return [torch.cat(mm_embeddings, dim=0)]
+        # One `EncoderGroup` per modality — matches how the parent's
+        # forward emits one row block per param and lets the framework
+        # do the per-modality reorder in one place.
+        def _encoder_fn(modality):
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Nit:

* `_make_encoder_fn`?
* You could consider having `_encoder_fn(multimodal_params, modality)` without the inner `_fn`, and pass `partial(_encoder_fn, modality="image")` in e.g. line 972.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 2]
```diff
@@ -1882,7 +1907,7 @@ def _text_level_video_replace(
 
     def _process_images(
         self, images: List[Image.Image | torch.Tensor], text_prompt: str
-    ) -> Tuple[Dict[str, Any], torch.Tensor]:
+    ) -> Tuple[Dict[str, Any], str]:
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1912]: Can you explain why it was necessary to switch from input ids (int tensor) to a text prompt? Does this change affect any consumer of this class's public methods?

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 3]
```diff
@@ -2580,6 +2562,11 @@ def _resample_audios(
     },
     placeholder_placement=MultimodalPlaceholderPlacement.BEFORE_TEXT,
     placeholders_separator="\n",
+    # STRING pre-inserts placeholders in `mm_item_order` order; the OPENAI
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Hm... I remember I specifically had to make this change to make the nemotron v3 outputs match that of vLLM. If you run the E2E accuracy tests, do they show the same scores (within noise) both before and after the change?

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py [hunk 4]
```diff
@@ -107,13 +107,16 @@ def _lengths_by_modality(
         # Raw-prompt entrypoints (non chat-parsing) do not attach a manifest,
         # so this is the single enforcement point that a >1-modality request
         # must carry `mm_item_order` to make prompt-order reordering possible.
-        present = [m for m in modalities if mp.multimodal_data.get(m) is not None]
+        # Check the *request-level* modality set so multi-group models (each
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_multimodal_mixin.py:126]: Nit: this multi-group / group / cross-group / group-local verbiage is throwing me for a loop. Can we give some examples in the code comment as to what each term means? Or, have it as a glossary at the class-level or module-level docstring.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 3 above)
Comment (aswinvisva) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: good catch - reverted, and updated the `mm_item_order` to have a model-specific override, so that the embeddings can be ordered in the way that nemotron expects using the same helpers

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 1 above)
Comment (aswinvisva) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Done both

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 2 above)
Comment (aswinvisva) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1912]: The reason was that before, the helper assumed there could be only one of images / videos / audio in the prompt , so it would apply that modality's placeholder expansion and tokenize directly inside the helper. Now that we may have all 3 modalities in the same request, each helper returns the expanded prompt so they can be composed one after the other on the running string, with a single final `tokenizer.encode` in 
`call_with_text_prompt` once all the necessary expansions have been applied.

This is just affecting the private functions: _process_images / _process_images_dynamic / _process_video_prompts / _process_audio are private helpers only called from call_with_text_prompt, so no public API or subclass hook is affected.

Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1912]: Great, thanks for the explanation. Might be worth leaving some breadcrumbs as code comments? Doesn't need to be super detailed, just enough for us to remember.

File: tensorrt_llm/inputs/registry.py [hunk 5]
```diff
@@ -715,12 +715,21 @@ class MultimodalPlaceholderMetadata:
             user's message.
             When False (default), placeholders are bulk-prepended or appended
             according to placeholder_placement.
+        - prompt_modality_order:
+            Fixed modality-major order the chat template emits placeholders in
+            when it regroups them (e.g. Nano's Jinja counts modalities and
+            emits image → video → audio regardless of chat-part send order).
+            When set, `MultimodalDataTracker.item_order()` returns items sorted
```
Comment (2ez4bz) [inline tensorrt_llm/inputs/registry.py:723]: TIL...

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py [hunk 6]
```diff
@@ -558,67 +558,6 @@ def _build_inputs(prompts_subset, media_subset):
         )
 
 
-class TestEncodeMultimodalDispatch:
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:567]: I'm assuming you did it already, but just in case not: might be worth double-checking we're not losing coverage by removing these tests.

File: tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+def _mp(*, image=None, video=None, audio=None):
+    data = {}
+    if image is not None:
+        data["image"] = image
+    if video is not None:
+        data["video"] = video
+    if audio is not None:
+        data["audio"] = audio
+    return SimpleNamespace(multimodal_data=data, mm_item_order=None)
```
Comment (2ez4bz) [inline tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py]: Nit: can we use the actual object here instead?

File: tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+from types import SimpleNamespace
+from unittest.mock import MagicMock
+
+import torch
+
+from tensorrt_llm._torch.models.modeling_nemotron_nano import NemotronH_Nano_VL_V2
+
+
+def _mp(*, image=None, video=None, audio=None):
```
Comment (2ez4bz) [inline tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py]: Nit: `_make_multimodal_params`?

**Review by 2ez4bz — APPROVED** [review]

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py (hunk 6 above)
Comment (aswinvisva) [reply tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:567]: Confirmed — the coverage previously in these classes is now split across two files:

* `_encode_multimodal` was a single orchestrator; it's now three per-modality methods, each unit-tested directly in `tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py`:
  * `test_property_registers_three_per_modality_groups` — group registration.
  * `test_image_group_wraps_virtual_params_and_concats` — image dispatch + virtual-param wrapping (equivalent to the old `test_encode_multimodal_dispatches_image`).
  * `test_audio_group_concats_encode_audio_outputs` — audio dispatch (equivalent to the old `test_encode_multimodal_dispatches_audio`).
  * `test_video_group_stashes_num_tokens_in_video` — EVS per-video count stash.
  * `test_video_group_skips_audio_when_video_has_no_audio` — replaces the old `test_video_without_audio_skips_audio_concat`.
  * `test_video_group_skips_audio_when_sound_encoder_missing` — replaces the old `test_no_audio_concat_when_sound_encoder_is_none`.
  * `test_video_group_interleaves_when_embedded_audio_present` — replaces the video-audio dispatch/orchestration part of `test_multi_video_audio_interleaving`.
* The interleave math (`_interleave_video_audio_embeddings`) is still directly tested by the untouched `TestInterleaveVideoAudioEmbeddings` in this same file — that's the numerical side of what `test_multi_video_audio_interleaving` covered.

The only tests I didn't map to a replacement are the four in `TestEncodeMultimodalContract` (`test_returns_list_with_a_single_element`, etc.) — the contract they tested is genuinely different now (`_encode_multimodal` returned `List[Tensor]`; the new `encode_multimodal_by_groups` returns a `Tensor` per the mixin), so those are obsolete rather than dropped.

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py [hunk 9]
```diff
@@ -1124,35 +851,56 @@ def _make_mock_model(self):
...
+        `multimodal_embedding_lengths` is required for the group encoder path:
+        `_lengths_by_modality` uses it to slice the encoder output back into
+        per-modality tensors. Single-item request → single-entry list.
+        """
         embed_mask_cumsum = torch.arange(1, num_tokens + 1, dtype=torch.int64)
         runtime = MultimodalRuntimeData(
             past_seen_token_num=0,
             chunk_end_pos=num_tokens,
             embed_mask_cumsum=embed_mask_cumsum,
         )
```
[the remark this answers is not in the document]
Comment (aswinvisva) [reply tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:875]: Fair callout — but the invariant these tests were actually written to protect is the `get_multimodal_embeddings` return-type contract that the encoder path assumes, not the chunked-prefill scheduler state machine. The regression risk this PR could plausibly introduce is: the new group encoder_fn returns something incompatible with `_cache_multimodal_embeddings`, and cache reuse silently breaks. That's what "second call skips encoder" is asserting.

Advancing `past_seen_token_num` would exercise a different property (correct partial-chunk masking), which is orthogonal to this refactor and better tested in the existing `test_multimodal_runtime.py` suite. Worth a follow-up if that coverage isn't there, but I'd rather keep this PR scoped.

File: tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                    video={
+                        "marker": 5,
+                        "rows": 4,
+                        "audio": {"has_audio": [True], "audio_num_clips": torch.tensor([1])},
+                    }
+                )
+            ]
+        )
+        m._encode_audio.assert_called_once()
+        m._interleave_video_audio_embeddings.assert_called_once()
```
[the remark this answers is not in the document]
Comment (aswinvisva) [reply tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py:172]: These files are already picked up by the pre-merge CI list glob at `tests/integration/test_lists/test-db/l0_h100.yml`:

```
- unittest/_torch/multimodal -k "not nemotron_nano_v2_vl_fp8"
```

That's a directory-level entry that pytest-collects everything under `tests/unittest/_torch/multimodal/`, including this new file — the tests I added ran under CI's `L0_MergeRequest_PR` last cycle without any list edit. No per-test entry needed.

File: tensorrt_llm/_torch/models/modeling_multimodal_mixin.py (hunk 4 above)
Comment (aswinvisva) [reply tensorrt_llm/_torch/models/modeling_multimodal_mixin.py:126]: Added the glossary to the code docstring

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 2 above)
Comment (aswinvisva) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py:1912]: added the code comments at the caller site

File: tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py (hunk 7 above)
Comment (aswinvisva) [reply tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py]: Done

File: tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py (hunk 8 above)
Comment (aswinvisva) [reply tests/unittest/_torch/multimodal/test_nemotron_nano_encoder_groups.py]: Done
